### ISSUE TITLE
About the usage of the loss function

### ISSUE THREAD

**Issue by Barry-Liang** [mk-minchul/AdaFace#7]
Hi, authors. Thanks for open sourcing such great work. Here I have a question about the usage of the loss function. In the readme [https://github.com/mk-minchul/AdaFace#usage](url), you take the embedding before normalization as input for the adaface loss. But in your implementation, [https://github.com/mk-minchul/AdaFace/blob/c4052220c51167a18c35ce15a450044180cbb281/train_val.py#L54](url) you take the embedding after normalization as input for the adaface loss. I am confused about that.  In addition, since you use the norm of the embedding feature as a metric of the image quality, do you think the normalization operation before generating the embedding feature will damage the efficiency of adaface loss. For example, CNN backbone->(1)L2 normalization->pooling->FC->embedding->(2)L2 normalization, will the first L2 normalization operation downgrade the efficiency of adaface loss?

**Comment by mk-minchul** [comment]
Dear Barry-Liang. Thanks for pointing it out the error. I have updated the readme to correctly reflect the usage. 
The input should be normalized to make it a cosine theta when it is multiplied with the normalized embedding.

As for the second question, my guess would be that if individual feature norm gets to be removed from the penultimate L2 norm layer, the subsequent FC layer might not be able to recover it sufficiently. Therefore, if normalization is needed, I would suggest using BatchNorm in (1) instead.

**Comment by Barry-Liang** [comment]
Thanks for your quick response. It makes sense.  I will close this issue.
